src/main/java/org/testng/internal/Parameters.java

```diff
@@ -264,35 +265,81 @@ else if (ITestResult.class.equals(c)) {
                                        MethodParameters params,
                                        XmlSuite xmlSuite) {
     List<Object> vResult = Lists.newArrayList();
-    for (int i = 0, j = 0; i < parameterTypes.length; i++) {
-      if (j < parameterNames.length) {
-        String p = parameterNames[j];
-        String value = params.xmlParameters.get(p);
-        if (null == value) {
-          // try SysEnv entries
-          value = System.getProperty(p);
+    if (isInjected) {
+      FilterOutInJectedTypesResult filterOutResult = FilterOutInJectedTypesFromOptionalValues(parameterTypes, optionalValues);
+      optionalValues = filterOutResult.getOptionalValues();
+      parameterTypes = filterOutResult.getParameterTypes();
+    }
+    for (int i = 0; i < parameterNames.length; i++) {
+      String p = parameterNames[i];
+      String value = params.xmlParameters.get(p);
+      if (null == value) {
+        // try SysEnv entries
+        value = System.getProperty(p);
+      }
+      if (null == value) {
+        if (optionalValues != null) {
+            value = optionalValues[i];
         }
         if (null == value) {
-          if (optionalValues != null) {
-            value = optionalValues[i];
-          }
-          if (null == value) {
             throw new TestNGException("Parameter '" + p + "' is required by "
                     + methodAnnotation
                     + " on " + prefix + " "
                     + name
                     + " but has not been marked @Optional or defined\n"
                     + (xmlSuite.getFileName() != null ? "in "
-                    + xmlSuite.getFileName() : ""));
-          }
+                            + xmlSuite.getFileName() : ""));
+        }
+    }
+      vResult.add(convertType(parameterTypes[i], value, p));
+    }
+    return vResult;
+  }
+  
+  /**
+   * Remove injected types from parameterTypes and optionalValues
+   * @param parameterTypes
+   * @param optionalValues
+   * @return FilterOutInJectedTypesResult
+   */
+  private static FilterOutInJectedTypesResult FilterOutInJectedTypesFromOptionalValues(Class<?>[] parameterTypes, String[] optionalValues) {
```

Comment:
Method names to begin with small letters.

Reply:
Also maybe we can have this method as a `package` visible method. That way we could add a unit test for this method as well by creating `ParametersTest`

Reply:
> Method names to begin with small letters.

Yes, I will fix it. 

> Also maybe we can have this method as a package visible method.

I prefer keeping modifier of this inner class as private . I truely don't want to explore any unnecessary access right outside, even within package.

Reply:
The intent is to ensure that we can enrich/enhance testability of this class. Rather than trying to figure out a complex test to get this codepath tested, why not expose this method at a package level, and add a test at the package level to test it out ? This class is anyway residing within the `internal` package, which kind of re-iterates to external consumers that this method is not meant to be used by external consumers and as such any changes can happen without prior notification.

Reply:
Although it violates code rule in my mind, I understand your intention. I will change it anyway.